apisix/plugins/ai-rag.lua

```diff
@@ -0,0 +1,168 @@
+local next = next
+local require = require
+
+local azure_ai_search_schema = {
+    type = "object",
+    properties = {
+        endpoint = {
+            type = "string",
+        },
+        api_key = {
+            type = "string",
+        },
+    }
+}
+
+local azure_openai_embeddings = {
+    type = "object",
+    properties = {
+        endpoint = {
+            type = "string",
+        },
+        api_key = {
+            type = "string",
+        },
+    },
+    required = { "endpoint", "api_key" }
+}
+
+
+local schema = {
+    type = "object",
+    properties = {
+        type = "object",
+        embeddings_provider = {
+            type = "object",
+            properties = {
+                azure_openai = azure_openai_embeddings
+            },
+            -- change to enum while implementing support for other search services
+            required = { "azure_openai" },
+        },
+        vector_search_provider = {
+            type = "object",
+            properties = {
+                azure_ai_search = azure_ai_search_schema
+            },
+            -- change to enum while implementing support for other search services
+            required = { "azure_ai_search" }
+        },
+    },
+    required = { "embeddings_provider", "vector_search_provider" }
+}
+
+local request_schema = {
+    type = "object",
+    properties = {
+        ai_rag = {
+            type = "object",
+            properties = {
+                vector_search = {},
+                embeddings = {},
+            },
+            required = { "vector_search", "embeddings" }
+        }
+    }
+}
+
+local _M = {
+    version = 0.1,
+    priority = 1060, -- TODO check with other ai plugins
+    name = "ai-rag",
+    schema = schema,
+}
+
+
+function _M.check_schema(conf)
+    return core.schema.check(schema, conf)
+end
+
+
+function _M.access(conf, ctx)
+    local httpc = http.new()
+    local body_tab, err = core.request.get_json_request_body_table()
+    if not body_tab then
+        return 400, err
+    end
+    if not body_tab["ai_rag"] then
+        core.log.error("request body must have \"ai-rag\" field")
+        return 400
+    end
+
+    local embeddings_provider = next(conf.embeddings_provider)
+    local embeddings_provider_conf = conf.embeddings_provider[next(conf.embeddings_provider)]
+    local embeddings_driver = require("apisix.plugins.ai-rag.embeddings." .. embeddings_provider)
+
+    local vector_search_provider = next(conf.vector_search_provider)
+    local vector_search_provider_conf = conf.vector_search_provider[vector_search_provider]
+    local vector_search_driver = require("apisix.plugins.ai-rag.vector-search." ..
+                                        vector_search_provider)
+
+    local vs_req_schema = vector_search_driver.request_schema
+    local emb_req_schema = embeddings_driver.request_schema
+
+    request_schema.properties.ai_rag.properties.vector_search = vs_req_schema
```

Comment:
Are there risks with module variables in concurrent scenarios?

Reply:
I don't think so, can you elaborate?

Reply:
I checked the process again and there is no risk of race condition. Ignore it.